crypto/ml_kem/ml_kem.c

```diff
@@ -1933,3 +2086,395 @@ int ossl_ml_kem_pubkey_cmp(const ML_KEM_KEY *key1, const ML_KEM_KEY *key2)
+                       key->vinfo->algorithm_name);
+        OPENSSL_free(*out);
+        return 0;
+    }
+
+    return (int) publen;
+}
+
+/* Allocate and encode PKCS#8 private key payload. */
+int ossl_ml_kem_i2d_prvkey(const ML_KEM_KEY *key, uint8_t **out,
+                           const char *formats)
+{
+    const ML_KEM_VINFO *v = key->vinfo;
+    ML_KEM_PKCS8_PREF *vp8_alloc, *vp8_slot;
+    const ML_KEM_PKCS8_INFO *vp8;
+    int len = ML_KEM_SEED_BYTES;
+    uint8_t *buf = NULL, *pos;
+    int ret = 0;
+
+    if (!ossl_ml_kem_have_prvkey(key)) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_NOT_A_PRIVATE_KEY,
+                       "no %s private key data available",
+                       key->vinfo->algorithm_name);
+        return 0;
+    }
+
+    vp8_slot = vp8_alloc = vp8_order(v->pkcs8_info, formats);
+    if (vp8_alloc == NULL)
+        return 0;
+
+    /* If we don't have a seed, skip seedful entries */
+    if (!ossl_ml_kem_have_seed(key))
+        while (vp8_slot->vp8_entry != NULL
+               && vp8_slot->vp8_entry->seed_offset != 0)
+            ++vp8_slot;
+    vp8 = vp8_slot->vp8_entry;
+    /* No matching table entries, give up */
+    if (vp8 == NULL || (len = (int) vp8->p8_bytes) == 0)
+        goto end;
+
+    if (out == NULL) {
+        ret = len;
+        goto end;
+    }
+
+    if ((pos = buf = OPENSSL_malloc((size_t) len)) == NULL)
+        goto end;
+
+    pos = OPENSSL_store_u32_be(pos, vp8->p8_magic);
+    if (vp8->seed_offset != 0) {
+        /*
+         * Either the tag/len were already included in |magic| or they require
+         * is to write two bytes now.
+         */
+        if (pos == buf + vp8->seed_offset - 2)
+            pos = OPENSSL_store_u16_be(pos, vp8->seed_magic);
+        else if (pos != buf + vp8->seed_offset)
+            goto end;
+        if (!ossl_ml_kem_encode_seed(pos, ML_KEM_SEED_BYTES, key))
+            goto end;
+        pos += ML_KEM_SEED_BYTES;
+    }
+    if (vp8->priv_offset != 0) {
+        if (pos == buf + vp8->priv_offset - 4)
+            pos = OPENSSL_store_u32_be(pos, vp8->priv_magic);
+        else if (pos != buf + vp8->priv_offset
+                 || vp8->priv_offset != sizeof(vp8->p8_magic)
+                 || vp8->priv_magic != vp8->p8_magic)
+            goto end;
+        if (!ossl_ml_kem_encode_private_key(pos, v->prvkey_bytes, key))
+            goto end;
+        pos += v->prvkey_bytes;
+    }
+    /* OQS form output with tacked-on public key */
+    if (vp8->pub_offset != 0) {
+        /* The OQS pubkey is not separately DER-wrapped */
+        if (pos != buf + vp8->pub_offset)
+            goto end;
+        if (!ossl_ml_kem_encode_public_key(pos, v->pubkey_bytes, key))
+            goto end;
+        pos += v->pubkey_bytes;
+    }
+
+    if (pos == buf + len) {
+        *out = buf;
+        ret = len;
+    }
+
+  end:
+    OPENSSL_free(vp8_alloc);
+    if (ret == 0)
+        OPENSSL_free(buf);
+    return ret;
+}
+
+# include <internal/encoder.h>
```

Comment:
I dislike includes in the middle of C source. Is there a strong reason for this?

Reply:
Unused in the FIPS module, specifically used only inthis one final function inside `#ifndef FIPS_MODULE`.
We don't have good tooling for detecting obsoleted `#includes`, that'd be nice to have.
At the top, should this be inside `#ifndef FIPS_MODULE` or just unconditional?

Reply:
I am not even sure putting `#ifndef FIPS_MODULE` around changes anything. The `internal/encoder.h` file is already part of the fips source checksums.

Also I've noticed - we use `"` instead of `<>` around internal headers such as those in internal and crypto subdirs. Not that it matters functionally but this convention is applied throughout the code.